src/components/Editor/BacklinkExtension.tsx

```diff
@@ -0,0 +1,177 @@
+import { Extension } from "@tiptap/core";
+import { Plugin, PluginKey, TextSelection } from "@tiptap/pm/state";
+import { Decoration, DecorationSet } from "@tiptap/pm/view";
+import { SuggestionsState } from "./BacklinkSuggestionsDisplay";
+
+const backlinkPlugin = (
+  updateSuggestionsState: (state: SuggestionsState | null) => void
+) => {
+  // Timeout to hide the suggestions after a certain amount of time. This is needed so that blur or focusout events don't hide the suggestions immediately.
+  let hideTimeout: NodeJS.Timeout | null = null;
+
+  return new Plugin({
+    key: new PluginKey("backlinks"),
+    state: {
+      init(_, { doc }) {
+        return DecorationSet.create(doc, []);
+      },
+      apply(tr, set) {
+        return set.map(tr.mapping, tr.doc);
+      },
+    },
+    view() {
+      return {
+        update: (view) => {
+          const { state } = view;
+          const { doc, selection } = state;
+          const { from } = selection;
+
+          if (hideTimeout) {
+            clearTimeout(hideTimeout);
+          }
+
+          if (!view.hasFocus()) {
+            hideTimeout = setTimeout(() => {
+              updateSuggestionsState(null);
+            }, 1000);
+            return;
+          }
+
+          const textBeforeCursor = doc.textBetween(0, from, "\n");
+          const lastOpeningBracketIndex = textBeforeCursor.lastIndexOf("[[");
+
+          if (
+            lastOpeningBracketIndex === -1 ||
+            textBeforeCursor.lastIndexOf("]]") > lastOpeningBracketIndex
+          ) {
+            updateSuggestionsState(null);
+            return;
+          }
+
+          const textToLeft = textBeforeCursor.slice(
+            lastOpeningBracketIndex + 2,
+            from
+          );
+
+          const coords = view.coordsAtPos(from);
+          updateSuggestionsState({
+            textWithinBrackets: textToLeft,
+            position: coords,
```

Comment:
Re usability nit - Perhaps the coords here should use the back ]] square brackets

Reply:
very fiddly to do this correctly - for now just added an offset to the coordinate